File: frontend/coordinator-android/app/src/main/java/ug/qaat/coordinator/ui/FreeRoomPicker.kt
package ug.qaat.coordinator.ui

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import ug.qaat.coordinator.net.FreeRoom
import ug.qaat.coordinator.net.FreeRoomsClient

/**
 * THE ROOM YOU CAN ACTUALLY USE.
 *
 * The timetabled room is locked, double-booked, being repainted, or an exam has it — and a class is
 * standing in the corridor. This lists every room in the institution that is free right now: all
 * colleges, all departments, all blocks, checked against the timetable AND against sessions other
 * coordinators have open at this second.
 *
 * WHY THE WHOLE INSTITUTION. The free room is usually somebody else's. A list bounded by the
 * coordinator's own department would hide exactly the rooms that are available, which is how this
 * ends with a class in a corridor while three empty halls sit one block away.
 *
 * PICKING ONE IS A DECLARATION, not a silent substitution. It marks the session as a provision, and
 * the QA monitors are told before they set off — because otherwise a monitor walks to the
 * timetabled room, finds it empty, and files "not taught" against a lecturer who is teaching thirty
 * metres away, and nobody can afterwards prove otherwise.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FreeRoomPicker(
    onDismiss: () -> Unit,
    onChosen: (room: FreeRoom, reason: String) -> Unit,
) {
    val scope = rememberCoroutineScope()
    var rooms by remember { mutableStateOf<List<FreeRoom>?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var onlyFree by remember { mutableStateOf(true) }
    var chosen by remember { mutableStateOf<FreeRoom?>(null) }
    var reason by remember { mutableStateOf("") }

    suspend fun load() {
        val token = AppState.token
        if (token == null) { error = "Not signed in"; rooms = emptyList(); return }
        val list = runCatching { FreeRoomsClient().list(token) }
            .getOrElse { error = "Could not check the rooms — you need signal for this."; emptyList() }
        rooms = list
        if (list.isEmpty() && error == null) error = "No rooms came back."
    }
    LaunchedEffect(Unit) { load() }

    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(bottom = 24.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Text("Find a free room", fontWeight = FontWeight.Bold, fontSize = 18.sp)
            Text(
                "Every room in the institution, checked against the timetable and against sessions " +
                    "running right now. Needs signal — a cached answer would send two classes to one room.",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            if (rooms == null) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    CircularProgressIndicator(Modifier.size(18.dp), strokeWidth = 2.dp)
                    Spacer(Modifier.width(10.dp)); Text("Checking every room…")
                }
            }
            error?.let { Text(it, color = MaterialTheme.colorScheme.error) }

            chosen?.let { r ->
                // Confirming is a second, deliberate step: the reason is the only account anyone
                // will ever have of WHY the lecture moved, and it is asked for while the
                // coordinator still remembers.
                Surface(
                    color = MaterialTheme.colorScheme.secondaryContainer,
                    shape = MaterialTheme.shapes.medium, modifier = Modifier.fillMaxWidth(),
                ) {
                    Column(Modifier.padding(14.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text("Use ${r.name}?", fontWeight = FontWeight.Bold)
                        Text(
                            "${r.building.ifBlank { "no block listed" }} · ${r.school.ifBlank { "no college" }}" +
                                if (r.capacity > 0) " · seats ${r.capacity}" else "",
                            style = MaterialTheme.typography.labelSmall,
                        )
                        OutlinedTextField(
                            value = reason, onValueChange = { reason = it },
                            label = { Text("Why the timetabled room could not be used") },
                            supportingText = { Text("Optional, but it is the only record of the reason.") },
                            modifier = Modifier.fillMaxWidth(),
                        )
                        Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                            Button(
                                modifier = Modifier.weight(1f),
                                onClick = { onChosen(r, reason.trim()); onDismiss() },
                            ) { Text("Use this room") }
                            OutlinedButton(
                                modifier = Modifier.weight(1f),
                                onClick = { chosen = null; reason = "" },
                            ) { Text("Back") }
                        }
                        Text(
                            "The QA monitors, the QA office and the lecturer are told immediately, " +
                                "so nobody visits the empty room.",
                            style = MaterialTheme.typography.labelSmall,
                        )
                    }
                }
            }

            if (chosen == null) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Checkbox(checked = onlyFree, onCheckedChange = { onlyFree = it })
                    Text("Free only")
                    Spacer(Modifier.weight(1f))
                    TextButton(onClick = { scope.launch { rooms = null; error = null; load() } }) { Text("Refresh") }
                }
                // Grouped by block: a coordinator with a waiting class cares about distance first.
                val list = (rooms ?: emptyList()).filter { !onlyFree || it.free }
                if (rooms != null && list.isEmpty()) {
                    Text(
                        "Nothing is free right now anywhere in the institution. Untick “Free only” " +
                            "to see what is holding each room, and until when.",
                        style = MaterialTheme.typography.bodySmall,
                    )
                }
                list.groupBy { it.building.ifBlank { "Unlisted block" } }.forEach { (block, inBlock) ->
                    Text(
                        "$block · ${inBlock.count { it.free }} free of ${inBlock.size}",
                        style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    inBlock.forEach { r ->
                        Surface(
                            shape = MaterialTheme.shapes.small, tonalElevation = 1.dp,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Row(
                                Modifier.padding(12.dp), verticalAlignment = Alignment.CenterVertically,
                                horizontalArrangement = Arrangement.spacedBy(10.dp),
                            ) {
                                Column(Modifier.weight(1f)) {
                                    Text(r.name, fontWeight = FontWeight.SemiBold)
                                    Text(
                                        buildString {
                                            append(r.school.ifBlank { "no college" })
                                            if (r.capacity > 0) append(" · seats ${r.capacity}")
                                            if (!r.free) {
                                                append(" · ${r.occupiedBy}")
                                                if (r.occupiedUntil.isNotBlank()) append(" until ${r.occupiedUntil}")
                                                if (r.occupiedNote.isNotBlank()) append(" — ${r.occupiedNote}")
                                            }
                                        },
                                        style = MaterialTheme.typography.labelSmall,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                                    )
                                }
                                if (r.free) {
                                    Button(onClick = { chosen = r }) { Text("Choose") }
                                } else {
                                    Text("IN USE", style = MaterialTheme.typography.labelSmall,
                                        color = MaterialTheme.colorScheme.error)
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: frontend/coordinator-flutter/lib/ui/coordinator/free_room_picker.dart
import 'package:flutter/material.dart';

import '../../net/coordinator_client.dart';

/// THE ROOM YOU CAN ACTUALLY USE — a faithful port of the native `FreeRoomPicker`.
///
/// Lists every room in the institution that is free right now (all colleges, all
/// departments, all blocks), checked against the timetable AND live sessions.
/// Picking one marks the session as a provision and the QA monitors are told
/// before they set off. Needs signal — a cached answer would send two classes to
/// one room.
Future<ProvisionalRoom?> showFreeRoomPicker(
  BuildContext context, {
  required String token,
  required CoordinatorClient client,
}) {
  return showModalBottomSheet<ProvisionalRoom>(
    context: context,
    isScrollControlled: true,
    builder: (_) => _FreeRoomPicker(token: token, client: client),
  );
}

/// The outcome when a coordinator commits to a room: the venue plus the reason.
class ProvisionalRoom {
  const ProvisionalRoom({required this.room, required this.reason});
  final FreeRoom room;
  final String reason;
}

class _FreeRoomPicker extends StatefulWidget {
  const _FreeRoomPicker({required this.token, required this.client});

  final String token;
  final CoordinatorClient client;

  @override
  State<_FreeRoomPicker> createState() => _FreeRoomPickerState();
}

class _FreeRoomPickerState extends State<_FreeRoomPicker> {
  List<FreeRoom>? _rooms;
  String? _error;
  var _onlyFree = true;
  FreeRoom? _chosen;
  String _reason = '';

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    setState(() {
      _rooms = null;
      _error = null;
    });
    if (widget.token.isEmpty) {
      setState(() {
        _error = 'Not signed in';
        _rooms = const [];
      });
      return;
    }
    List<FreeRoom> list;
    try {
      list = await widget.client.freeRooms(widget.token);
    } on Object {
      list = const [];
      if (mounted) {
        setState(() => _error =
            'Could not check the rooms — you need signal for this.');
      }
    }
    if (!mounted) return;
    setState(() {
      _rooms = list;
      if (list.isEmpty && _error == null) _error = 'No rooms came back.';
    });
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final scheme = theme.colorScheme;
    return Padding(
      padding: EdgeInsets.only(
        left: 20,
        right: 20,
        bottom: 24 + MediaQuery.of(context).viewInsets.bottom,
      ),
      child: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            const Text(
              'Find a free room',
              style: TextStyle(fontWeight: FontWeight.bold, fontSize: 18),
            ),
            const SizedBox(height: 4),
            Text(
              'Every room in the institution, checked against the timetable and '
              'against sessions running right now. Needs signal — a cached answer '
              'would send two classes to one room.',
              style: theme.textTheme.labelMedium?.copyWith(
                color: scheme.onSurfaceVariant,
              ),
            ),
            const SizedBox(height: 10),

            if (_rooms == null)
              Row(
                children: [
                  const SizedBox(
                    width: 18,
                    height: 18,
                    child: CircularProgressIndicator(strokeWidth: 2),
                  ),
                  const SizedBox(width: 10),
                  const Text('Checking every room…'),
                ],
              ),
            if (_error != null)
              Text(_error!, style: TextStyle(color: scheme.error)),
            const SizedBox(height: 6),

            if (_chosen != null) _confirmCard(scheme) else _listCard(scheme),
          ],
        ),
      ),
    );
  }

  Widget _confirmCard(ColorScheme scheme) {
    final r = _chosen!;
    return Card(
      color: scheme.secondaryContainer,
      child: Padding(
        padding: const EdgeInsets.all(14),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text('Use ${r.name}?', style: const TextStyle(fontWeight: FontWeight.bold)),
            const SizedBox(height: 6),
            Text(
              '${r.building.isEmpty ? 'no block listed' : r.building} · '
              '${r.school.isEmpty ? 'no college' : r.school}'
              '${r.capacity > 0 ? ' · seats ${r.capacity}' : ''}',
              style: Theme.of(context).textTheme.labelSmall,
            ),
            const SizedBox(height: 8),
            TextField(
              onChanged: (v) => setState(() => _reason = v),
              decoration: const InputDecoration(
                labelText: 'Why the timetabled room could not be used',
                helperText: 'Optional, but it is the only record of the reason.',
                border: OutlineInputBorder(),
              ),
            ),
            const SizedBox(height: 10),
            Row(
              children: [
                Expanded(
                  child: FilledButton(
                    onPressed: () => Navigator.of(context).pop(
                      ProvisionalRoom(room: r, reason: _reason.trim()),
                    ),
                    child: const Text('Use this room'),
                  ),
                ),
                const SizedBox(width: 10),
                Expanded(
                  child: OutlinedButton(
                    onPressed: () => setState(() {
                      _chosen = null;
                      _reason = '';
                    }),
                    child: const Text('Back'),
                  ),
                ),
              ],
            ),
            const SizedBox(height: 6),
            Text(
              'The QA monitors, the QA office and the lecturer are told '
              'immediately, so nobody visits the empty room.',
              style: Theme.of(context).textTheme.labelSmall,
            ),
          ],
        ),
      ),
    );
  }

  Widget _listCard(ColorScheme scheme) {
    final theme = Theme.of(context);
    final list = (_rooms ?? const <FreeRoom>[])
        .where((r) => !_onlyFree || r.free)
        .toList();
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Row(
          children: [
            Checkbox(value: _onlyFree, onChanged: (v) => setState(() => _onlyFree = v ?? true)),
            const Text('Free only'),
            const Spacer(),
            TextButton(onPressed: _load, child: const Text('Refresh')),
          ],
        ),
        if (_rooms != null && list.isEmpty)
          Text(
            'Nothing is free right now anywhere in the institution. Untick '
            '"Free only" to see what is holding each room, and until when.',
            style: theme.textTheme.bodySmall,
          ),
        for (final MapEntry(key: block, value: inBlock)
            in _byBlock(list).entries) ...[
          Text(
            '$block · ${inBlock.where((r) => r.free).length} free of '
            '${inBlock.length}',
            style: theme.textTheme.labelMedium?.copyWith(
              color: scheme.onSurfaceVariant,
            ),
          ),
          const SizedBox(height: 4),
          for (final r in inBlock)
            Card(
              elevation: 1,
              child: Padding(
                padding: const EdgeInsets.all(12),
                child: Row(
                  children: [
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(r.name, style: const TextStyle(fontWeight: FontWeight.w600)),
                          Text(
                            '${r.school.isEmpty ? 'no college' : r.school}'
                            '${r.capacity > 0 ? ' · seats ${r.capacity}' : ''}'
                            '${r.free ? '' : ' · ${r.occupiedBy}'
                                '${r.occupiedUntil.isNotEmpty ? ' until ${r.occupiedUntil}' : ''}'
                                '${r.occupiedNote.isNotEmpty ? ' — ${r.occupiedNote}' : ''}'}',
                            style: theme.textTheme.labelSmall?.copyWith(
                              color: scheme.onSurfaceVariant,
                            ),
                          ),
                        ],
                      ),
                    ),
                    if (r.free)
                      FilledButton(
                        onPressed: () => setState(() => _chosen = r),
                        child: const Text('Choose'),
                      )
                    else
                      Text(
                        'IN USE',
                        style: theme.textTheme.labelSmall?.copyWith(color: scheme.error),
                      ),
                  ],
                ),
              ),
            ),
        ],
      ],
    );
  }

  Map<String, List<FreeRoom>> _byBlock(List<FreeRoom> rooms) {
    final grouped = <String, List<FreeRoom>>{};
    for (final r in rooms) {
      grouped.putIfAbsent(r.building.isEmpty ? 'Unlisted block' : r.building, () => []).add(r);
    }
    return grouped;
  }
}